Enable integration with pre-commit

JohnPaton opened the pull request:
This is a first attempt at an implementation of the suggestion in #691 to enable "proper" integration with the pre-commit framework. The changes summarized:
* A new `.pre-commit-hooks.yaml` file, which tells `pre-commit` how to install this package (pre-commit does this internally), which command to run, and what the default options are
* A new cli argument, `--add-untracked`, which adds any output files that weren't previously tracked by `git` to the index.  Two points to note here:
    * The reason we are tracking only new files is that `pre-commit` doesn't consider changes to them if they aren't in the index (pre-commit requires all files in the index be unchanged by all hooks before passing)
    * The command will return a non-zero status code in case any newly generated files are actually added. This is because by adding them "on the fly" during the hook, pre-commit doesn't pass them to any other configured hooks for further checks, so they could be wrongly accepted into the commit.

The net effect is a proper pre-commit hook, which will only allow a commit as long as the configured Jupytext command doesn't change any of the files in the index, or add any new ones. This is how pre-commit's hooks are expected to operate.

Questions:
* This is a bit of an adjustment to how Jupytext's main command is operating - is it too much? How easy would it be to make this separate/not expose the `--add-untracked` option so publicly?
* How to test this behaviour (will be investigating, this is a WIP)?

Skylion007 commented:
I was actually just about to make this PR myself, @JohnPatton. The untracked feature is nice, but isn't it similar to the `--precommit` arg already?

Skylion007 commented:
The proper way of doing this pre-commit hook is to allow pre-commit to pass in the args directly into jupytext by using pass_filenames: True. The only thing that may need to be improved is adding in a filtering that automatically adds relvant RST, py or other supported filetypes (currently that has to be specified if it's not ipynb).

Skylion007 commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,10 @@
+  language: python
+  entry: jupytext --add-untracked
+  pass_filenames: true
+  files: ipynb
```
What about the other use case of generating ipynb from newly added Python files?

Skylion007 replied:
This should be use the type: precommit value instead of the files. The former is much more robust and is the recommended way to make notebook pre-commit hooks.

JohnPaton replied:
yep we already discussed the defaults in the issue, default behavior needs to be removed entirely. We do need a flag to ignore unused inputs instead of the current erroring.

Skylion007 commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,10 @@
+- id: jupytext
+  name: jupytext
+  description: Convert all notebooks into another format
```
This description isn't quite right if you want to the opposite to occur, (generate clean ipynb from python files).

JohnPaton replied:
Agreed, suggestions welcome

Skylion007 commented on jupytext/cli.py:
```diff
@@ -681,6 +689,9 @@ def jupytext_single_file(nb_file, args, log):
         write(notebook, nb_dest, fmt=dest_fmt)
         if args.pre_commit:
             system("git", "add", nb_dest)
+        elif args.add_untracked and is_untracked(nb_dest):
```
What about other untracked filetypes? Using `--sync` should handle this nicely. Doesn't pre-commit hooks really shouldn't call git add inside of themselves like this. It's against the standard pattern.

Skylion007 replied:
This is really bad actually because you could add files the are in the pre-commit global or hook-wide exclude . That would be very undesirable behavior.

JohnPaton replied:
Yep, will remove this in favour of just exiting non-zero if output is untracked (don't add)

Skylion007 commented on docs/using-pre-commit.md:
```diff
@@ -27,35 +27,41 @@ Note that these hooks do not update the `.ipynb` notebook when you pull. Make su
+    rev: master
+    hooks:
+    - id: jupytext
+      args: [--to, py:percent, --pipe, black]
```
Args should be none, (except maybe --sync). Also not all projects use black. These args should be flexible based on what the end user wants to do with it.

JohnPaton replied:
Because I wasn't aware of it :) From discussion with @mwouts in the issue it sounds like best option is not to provide any default args and force the user to always choose

Skylion007 commented on docs/using-pre-commit.md:
```diff
@@ -27,35 +27,41 @@ Note that these hooks do not update the `.ipynb` notebook when you pull. Make su
+    hooks:
+    - id: jupytext
+      args: [--to, py:percent, --pipe, black]
+-   repo: https://github.com/psf/black
```
This isn't necessary or dieal and could conflict with otehr black hooks. It should just be added as an `additional_dependency` but the end user if needed at all.

JohnPaton replied:
It's just an informational example

Skylion007 replied:
Oh this is abusing the virtualenv of the pre-commit hook (it should only require the bear minimum requirements and other hooks if they absolutely build upon one another). It should really be installed using additional_depenencies if needed by the end user (see the blog post test I adjusted.

JohnPaton replied:
Added black in additional_depenencies

mwouts replied:
Yes I agree with `additional_dependencies` (I think there is a typo in `additional_depenencies`).

Also you say _you must ensure that all hooks will pass (...)_. 
I understand that we are obliged to do so when we use `--to`, but maybe if we use `--sync` it's just "more convenient" as we need to run the hook just once, not twice. Is that correct?

mwouts replied:
@JohnPaton , what would you think of removing this example? I was trying to add a test for it, but we seem to have tricky issues - maybe we add that in another PR?
```
@requires_pre_commit
def test_pre_commit_hook_sync_black(tmpdir):
    # get the path and revision of this repo, to use with pre-commit
    repo_root = str(Path(__file__).parent.parent.resolve())
    repo_rev = system("git", "rev-parse", "HEAD", cwd=repo_root).strip()

    git = git_in_tmpdir(tmpdir)

    # set up the tmpdir repo with pre-commit
    pre_commit_config_yaml = dedent(
        f"""
        repos:
        - repo: {repo_root}
          rev: {repo_rev}
          hooks:
          - id: jupytext
            args: [--sync, --pipe, black]
            additional_dependencies:
              - black==19.10b0 # Matches hook
        
        - repo: https://github.com/psf/black
          rev: 19.10b0
          hooks:
          - id: black
            language_version: python3
        """
    )
    tmpdir.join(".pre-commit-config.yaml").write(pre_commit_config_yaml)
    git("add", ".pre-commit-config.yaml")
    with tmpdir.as_cwd():
        pre_commit(["install", "--install-hooks"])

    # write test notebook and output file
    nb = new_notebook(
        cells=[new_code_cell("1+1")],
        metadata={
            "jupytext": {"formats": "ipynb,py:percent", "main_language": "python"}
        },
    )
    nb_file = tmpdir.join("test.ipynb")
    py_file = tmpdir.join("test.py")
    write(nb, str(nb_file))

    git("add", ".")
    # First attempt to commit fails with message
    # "Output file test.py is not tracked in the git index"
    with pytest.raises(SystemExit):
        git("commit", "-m", "fails")

    git("commit", "-m", "succeeds")

    assert "test.ipynb" in git("ls-files")
    assert "test.py" in git("ls-files")
    # py_file should have been reformated by black
    assert "\n1 + 1\n" in py_file.read_text()
```

Skylion007 commented on docs/using-pre-commit.md:
````diff
@@ -27,35 +27,41 @@ Note that these hooks do not update the `.ipynb` notebook when you pull. Make su
 
-Using Jupytext with the [pre-commit package manager](https://pre-commit.com/) is another option. You could add the following to your `.pre-commit-config.yaml` file:
-```
+Using Jupytext with the [pre-commit package manager](https://pre-commit.com/) is another option. You could add the following to your `.pre-commit-config.yaml` file to convert all staged notebooks to python scripts in `py:percent` format (the default):
````
There is a use case where the old way of doing it actually preferred, and that's when you want to run locally installed code. This commit hook is great for a general use case where conversion is the only part of JuPyText you want to use, but execution requires being in the local sandbox.

JohnPaton replied:
I'm not sure I understand this comment. Why does it matter if it's installed locally or in the sandbox for this case?

Skylion007 replied:
Requirements that can't be specified by pip install or are difficult to (like say the local master version of a pip package).

Skylion007 commented on docs/using-pre-commit.md:
```diff
@@ -27,35 +27,41 @@ Note that these hooks do not update the `.ipynb` notebook when you pull. Make su
 repos:
--   repo: local
+-   repo: https://github.com/mwouts/jupytext
+    rev: master
```
This really should be pinned to a specific version or left for the user to specify. Otherwise it can cause a bunch of issues.

JohnPaton replied:
Again, this is an example.

Skylion007 reviewed (changes requested):
Great job on the test suite, but ultimately I think this pre-commit hook is too narrow/specific and doesn't address other common use cases of JuPyText (like generating Google Colab notebooks from scripts). 

@JohnPaton Happy to make the changes myself on this PR if given write access.

Skylion007 commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,10 @@
+  language: python
+  entry: jupytext --add-untracked
+  pass_filenames: true
+  files: ipynb
```
Instead of files why not use `types: [jupyter]` like nbstripout uses?

JohnPaton replied:
It didn't work on my machine, probably I had an older pre-commit installed but I wasn't sure when it was added so I did it this way. But we will remove this and just pass all files, ignoring the ones that we aren't trying to process.

Skylion007 replied:
@JohnPaton, yeah I've run into this before. You need to run `pip install -U identify`. Annoying that you can specify a min version of identify in the hook (you can for pre-commit though), but it is the recommended way to do it and will make it a lot easier for end users to override the behavior and support other types.
```suggestion
  types: [notebook, python]
```
We can also add markdown, etc to the list of types:

Skylion007 commented on docs/using-pre-commit.md:
```diff
@@ -27,35 +27,41 @@ Note that these hooks do not update the `.ipynb` notebook when you pull. Make su
-        language: system
+    - id: jupytext
+      args: [--to, py:light]
+    - id: jupytext
```
Why not have one hook id with `--to py:light --to markdown` it's much more efficient

JohnPaton replied:
Because I didn't know Jupytext supported that syntax, thanks :)

Skylion007 commented on tests/test_pre_commit.py:
```diff
@@ -364,3 +372,171 @@ def test_wrap_markdown_cell(tmpdir):
+
+
+def test_add_untracked_adds_new(tmpdir):
+    git = git_in_tmpdir(tmpdir)
```
with so many calls to git like this, why not just gitpython or some other nice bindings. :)

JohnPaton replied:
Agreed, I was just using the style that was already there, I'm not here to refactor the test suite

Skylion007 replied:
Ah, note to @mwouts then.

mwouts replied:
Oh yes I'm sorry I'm just discovering `gitpython`... I'll refactor that part later on :smile:

mwouts commented:
Hi @Skylion007 , thanks for joining the conversation! Well with @JohnPaton and yourself in the thread I am going to learn a LOT about pre-commit hooks, I am looking forward to it!

I don't know what you will think about that, but... I'd like to make the choices about this hook in a test-driven manner. I mean, I am thinking of a series of use cases that I'd like to see supported, like
1. all the notebooks in the repo are paired to `py:percent` files, and the `.ipynb` files are in `.gitignore`, so they don't reach the repository (the hook updates the `py` file from the `.ipynb` file that the user may be modifiying outside of Jupyter)
2. the notebooks are paired to `py:percent` files. The `.ipynb` files are committed with no outputs, but still the user can keep the outputs locally. And the hook ensures that modifications in whatever file are reported to the other one
3. the notebooks are paired to `py:percent` files, and synched with the `.ipynb` files. If a notebook was modified it is re-executed before the commit (I'm not so sure that this case makes sense - or do we want to reexecute the notebook only if the text version was changed?)

In addition to these use cases, I'd like to document and test (IMO the main purpose of writing the blog is to test that this is going to be usable :smile:) how to
- store the text versions and ipynb files in distinct folders (probably not related to the hook, this will directly depend on Jupytext)
- make sure that the `py` versions of the notebooks meets quality requirements (flake8) or formatting (black) - maybe this is not for our hook, but for a standard hook on `py` files
- and how to reformat the long lines in markdown cells into shorter ones, compliant with pep8

Which are your own expectations? How big is the overlap with mine?

Regarding the timing, I hope to have a draft for the blog by tomorrow or Friday, I'll share it with you both then, and then I hope to be able to iterate quite fast, within one or two weeks I mean.

Skylion007 commented:
In regards to three use cases:

1. This is fine as long as you don't modify the git staging. We could add an optional flag to throw errors if one of the edited files are not added to the index, instead of throwing it..  It doesn't matter if a file is in a gitignore if the file is added / currently staged for the commit.
1. Is actually very similar to what we do in habitat-sim and is accomplished by having two hooks (nbstripout being a seperate one to ensure all committed notebooks have no outputs). It won't edit the output of local files unless you try to commit them and those changes can be easily reverted / fixed with git pop/stash.
1. Is easy with default pre-commit settings, if you have --sync set, you can also have --execute. pre-commit will only execute on files that staged and changed in commit (unless manually run with --all-files).

By default, pre-commit hooks will only run on edited files for speed. Having it run on all files / the entire configuration is easily doable though.

* Yeah not related to the hook. Just the args that can be passed into the hook / the jupytext config.
* Easily done in both actually, with pipe + additional_requirements or by just editing the python files. I actually run them on both, but really only the editing the text files is necessary (requires two pre-commit runs in the later case).
* This seems unrelated and be can handled with pipe / additional_depenencies IMO. 

I'd say the overlap is quite large.

mwouts commented:
John, Aaron, I've invited you to https://github.com/mwouts/pre_commit_hooks_for_jupyter_notebooks.
It is very much a work in progress, though.

The first thing that I'd like to chat about is this "test" (WIP as well):
https://github.com/mwouts/pre_commit_hooks_for_jupyter_notebooks/blob/ac60bcbcec5ff65c171f8bbb03cc13a795e88a7d/examples/py_percent_sync_with_ipynb_with_outputs/test_py_percent_sync_with_ipynb_with_outputs.py#L35-L64
```
def test_py_percent_sync_with_ipynb_with_outputs(tmpdir, capsys):
    # Create an empty repository and initialize it
    repo = Repo.init(tmpdir)

    # Add the .pre-commit-config.yaml that corresponds to this example
    create_pre_commit_config_yaml_file(tmpdir, __file__)
    repo.index.add([".pre-commit-config.yaml"])
    repo.index.commit("Add .pre-commit-config.yaml")

    # Install pre-commit
    with working_directory(tmpdir):
        pre_commit(["install"])

    # Add a .jupytext.toml file that pairs all notebooks to py:percent files
    tmpdir.join(".jupytext.toml").write_text(
        "default_jupytext_formats = ipynb,py:percent"
    )
    repo.index.add([".jupytext.toml"])
    repo.index.commit("Add .jupytext.toml")

    # Add an ipynb file
    # Try to commit => pre-commit should create the py file, and fail!
    # Try to commit => pre-commit should complain that the py file is not in the index
    # Add the py file to the index
    # And commit (OK)

    # Modify the py file
    # Try to commit => pre-commit should complain that the ipynb file is not in the index
    # Add the ipynb file
    # And commit (OK)
```

I like the idea to use such a test to document explicitly our expectations. But at this stage I am not sure that `pre_commit(["install"])` in the `tmpdir` works. Then, do my comments match what we said on the hook? Do you deem this as user friendly (I mean, failing often and requiring 'git add')?

Skylion007 commented:
@mwouts Should John and I edited the repo directly? Left some comments on the questions in the readme.

Skylion007 commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,10 @@
+- id: jupytext
```
We need a copy of this hook using the system language so that it runs off the system version of JuPytext, happy to write it @JohnPaton . We can even reuse the node by taking advantage of some nice advanced YAML features.

JohnPaton replied:
Why do we need that? Users can just do (incomplete example)
```yaml 
- repo: local
  hooks:
    language: system
    entry: jupytext
```
if they want that

Skylion007 commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,10 @@
+- id: jupytext
+  name: jupytext
+  description: Convert all notebooks into another format
```
```suggestion
  description: Synchronizes all jupytext notebooks and paired files.
```

Skylion007 reviewed (changes requested):
Made some suggestions

JohnPaton commented:
I've given you write access to the PR @Skylion007 - I've got a pretty busy few days ahead so I'll be a bit less active than I'd like, sorry in advance

JohnPaton commented on jupytext/cli.py:
```diff
@@ -417,15 +429,24 @@ def jupytext_single_file(nb_file, args, log):
-            else full_path(base_path(nb_file, args.input_format), args.output_format)
-        )
-    )
+    nb_dest = None
+    if args.output:
+        nb_dest = args.output
+    elif args.output_format and nb_file == "-":
+        nb_dest = "-"
+    elif args.output_format:
+        try:
+            bp = base_path(nb_file, args.input_format)
+        except InconsistentPath:
+            if args.ignore_unmatched:
+                log(
+                    "[jupytext] Ignoring unmatched input "
+                    "path {} for format {}".format(nb_file, args.input_format)
+                )
+                return 0
+            else:
+                raise
+        nb_dest = full_path(bp, args.output_format)
```
Refactored the logic here a bit to make it clearer (to me at least) what's going on

JohnPaton commented:
Okay @mwouts @Skylion007, I've made some pretty big changes based on all the feedback (really feeling this collab guys, nice working with you!)

I removed `--add-untracked`, it's indeed wrong to modify the index. Instead, I've added two new options: 
* `--ignore-unmatched` so that `pre-commit` can pass the filenames and we can just ignore them if they don't match what we are trying to do
* and `--alert-untracked` which exits non-zero if the output files aren't tracked in the git index. That way we can always be sure that if the hook is making changes, either:
    * the files are tracked so `pre-commit` will fail the hook if they are modified, or 
    * the files are untracked, and `--alert-untracked` will fail the hook. 
   
  In each case the changes have to be added to the index before the hook can succeed

I have also removed default arguments beyond those two from the hook configuration, so that the user is forced to specify what they are trying to do. 

I still need to update the docs but wanted your feedback :)

JohnPaton commented:
codecov is complaining, but I'm not sure why

Skylion007 commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,6 @@
+  name: jupytext
+  description: Runs jupytext on all notebooks and paired files.
+  language: python
+  entry: jupytext --ignore-unmatched --alert-untracked
```
Should this really be in entry or passed args. The user should be able to override this if they really want without modifying the entry, right? Thoughts @mwouts ?

Skylion007 replied:
Note to self: Add proper types for file matching.

JohnPaton replied:
No, these are both essential to the hook functioning imo. Other rags can be passed but it has to consistently fail if changes are made, and it should ignore input files that it's not trying to do anything with. Users can specify types/files if they want.

Skylion007 commented on docs/using-pre-commit.md:
```diff
@@ -27,35 +27,41 @@ Note that these hooks do not update the `.ipynb` notebook when you pull. Make su
+repos:
+-   repo: https://github.com/mwouts/jupytext
+    rev: master
+    additional_depenencies: [black]
```
Stylistically, these usually go at the end of the hook. They can also use the yaml list syntax (enumerated with -black).

mwouts commented:
Hi @JohnPaton , @Skylion007 , thank you so much for your work here, I also enjoy very much this collaboration!

I completely agree with the `--alert-untracked` argument, you've convinced me that it is a better idea than the previous `--add-untracked`.

Regarding the `--ignore-unmatched`, I am not so sure that it is necessary. I expect that `jupytext --sync filename.ext` does not do anything if filename.ext is not the name of a paired notebook (well it reads the file and then issues a warning, but no error). It only fails if `ext` is not a notebook extension (i.e. something else than `ipynb`, `md`, `py`, 'R', etc).

Also, while trying the hook I saw a potential issue with the new build system for Jupytext - @JohnPaton you might not have seen it before because it was introduced just last week. I'll report on that at #706 .

JohnPaton commented:
I also hit that build issue now @mwouts because I rebased 😬

JohnPaton commented:
> I expect that jupytext --sync filename.ext does not do anything if filename.ext is not the name of a paired notebook

Hmmm, but if they are not using `--sync`, but `--from --to`, then it does throw errors

mwouts commented:
> Hmmm, but if they are not using --sync, but --from --to, then it does throw errors

Oh yes that's right... By the way I wanted to ask why you don't add `--sync` to the defaut command, but I think your comment answers my question, you want to run `--to` in some cases.

Regarding the build issue, I'll try to address it tonight. Meanwhile a quick fix would be to comment out the part of `setup.py` that depends on `jupyter-packaging` - you won't need that part of the package (the extension for Jupyter Lab) in the pre-commit hook anyway.

Skylion007 commented on .pre-commit-hooks.yaml:
```diff
@@ -0,0 +1,6 @@
+  description: Runs jupytext on all notebooks and paired files.
+  language: python
+  entry: jupytext --ignore-unmatched --alert-untracked
+  pass_filenames: true
```
@mwouts There is a possible race condition here, maybe? with `require_serial: False` it's possible that `jupytext -sync` is called on both paired files simultaneously in parallel. In practice, this shouldn't cause too many issues, but there could be edge cases I am not considering. `require_serial: True` would prevent the hook from running in parallel, but is probably safer behavior.

JohnPaton replied:
Great point!

mwouts replied:
Well I must admit that I never thought of that... but you're right, if one instance of Jupytext writes the result at the same time than the other one is starting to read them we will probably run into trouble... So yes I am in favor of `require_serial: True`

mwouts commented on docs/using-pre-commit.md:
```diff
@@ -27,35 +27,43 @@ Note that these hooks do not update the `.ipynb` notebook when you pull. Make su
+    rev: #CURRENT_TAG/COMMIT_HASH
+    hooks:
+    - id: jupytext
+      args: [--from, ipynb, --to, py:light, --to, markdown]
```
Oh sorry there can only be one value for `--to`

Skylion007 replied:
Ah my bad for that suggestion.

mwouts replied:
We still have to update the doc - would you like to take https://github.com/mwouts/jupytext/commit/07b323b92da78d24b7f6c0887adc8f9d30607f8f?

mwouts commented on tests/test_pre_commit.py:
```diff
@@ -364,3 +372,183 @@ def test_wrap_markdown_cell(tmpdir):
+    pre_commit_config_yaml = dedent(
+        f"""
+        repos:
+        - repo: {repo_root}
+          rev: {repo_rev}
```
Out of curiosity, is this a way to test the local version of the Jupytext hook? Well done!

Skylion007 replied:
The local repo hook can be used for this: https://pre-commit.com/#advanced

JohnPaton replied:
Yes it's to test the "current version", pre-commit will create a new env for it. So you know that what you are committing works

mwouts commented on jupytext/cli.py:
```diff
@@ -711,7 +732,11 @@ def write_function(path, fmt):
 
-    return 0
+    if args.alert_untracked and is_untracked(nb_dest):
+        log("[jupytext] Output file {} is not tracked in the git index".format(nb_dest))
```
This message is going to be seen lots of times. Maybe we should say explicitly that the user needs to call "git add" on that file?

Skylion007 replied:
Actually, it won't be seen at all unless the pre-commit hook is in verbose mode...

mwouts commented:
Hi @JohnPaton , @Skylion007 , well the PR is going well, isn't it? I think we've done most of the review, and must say that I like very much the contribution, and also the tests that you have added here.

Maybe there are two more small points that I'd like to see with you:
- I think that the error message when one file created by Jupytext is untracked is super important. Can we make it very clear that the user should simply run `git add` on that file?
- Also, the tests are for `--from ipynb --to py`, and that is a very nice use case, but I think that `--sync` is going to be more idiomatic. Do you have reasons to prefer the first form? Also, would you mind adding one test for that mode? Note that you will need to first turn the `py` or `ipynb` file into a paired notebook, which you can do with either a `jupytext.toml` file with this content: `"default_jupytext_formats = ipynb,py:percent"`, or with a call to `jupytext --set-formats ipynb,py:percent` on that file.

Also this is the first time I see a PR with a "draft" status, who is supposed to change it to "ready", is that you, @JohnPaton? On my side I'll continue working on the blog post but maybe we don't need to further postpone this PR, with the tests I think that we can have a good confidence that this is going to work well!

 Another question, probably I should plan to tag/release a new version when we integrate this?

JohnPaton commented:
> Can we make it very clear that the user should simply run git add on that file?

Agreed, done, let me know if this is good enough!

> Do you have reasons to prefer the first form?

Because my goal in using Jupytext is to get rid of Notebooks from our VCS entirely, even if users start their work by creating them. But I can switch one of the tests, sure! Unfortunately I can't verify that it works because I can no longer install Jupytext (blocked by #706), so my ability to contribute much more is a bit limited. 

> who is supposed to change it to "ready"

I'll do that now!

> Another question, probably I should plan to tag/release a new version when we integrate this?

I think so, we are introducing some new CLI args so probably a good idea

mwouts commented:
Hi @JohnPaton , I did some testing, and I think we need a few more updates to make sure that the `--sync` mode works well.

Also, I've started adding a few more commits on top of yours at https://github.com/mwouts/jupytext/commits/JohnPaton-pre-commit-hooks-2, but now the commit does not fail, I guess that I have not setup `types` correctly - do I need to add them to the `.pre-commit-config.yaml` as well?

> Because my goal in using Jupytext is to get rid of Notebooks from our VCS entirely
Ha ha! Well that is a point that I'd like to chat a bit more about... For instance, do you plan to remove the `ipynb` file from the index? In the documentation of the other implementation of pre-commit hook, we do have a line that says that one can run `git reset HEAD **/*.ipynb` at the end of the hook. Do you plan to do something similar? By the way, let me recall that a simple approach for that is to use _paired notebooks_ (the synchronisation is done by Jupyter itself) and then just add `*.ipynb` to `.gitignore`.

JohnPaton commented:
Hi @mwouts, I've rebased again with your changes and now it installs (yay!), but I can't run the tests still. Any ideas?

```console
$ pytest tests/test_pre_commit.py
ImportError while loading conftest '/Users/cw0504/personal/jupytext/tests/conftest.py'.
tests/conftest.py:15: in <module>
    jupytext.config.JUPYTEXT_CEILING_DIRECTORIES = ["/tmp/"]
E   AttributeError: module 'jupytext' has no attribute 'config'
exit 4
```

mwouts commented:
Hi @JohnPaton , I'm so sorry for this extra difficulty.

Locally I can't reproduce the `AttributeError: module 'jupytext' has no attribute 'config'` error.
Thus I would advice to recreate your dev environment from scratch, following [these instructions](https://github.com/mwouts/jupytext/blob/master/CONTRIBUTING.md#jupytext-as-a-python-package), i.e.
```
conda env create --file environment.yml  # or conda env update --file ...
conda activate jupytext-dev
python -m ipykernel install --name jupytext-dev --user
pip install -e .
```

Also, you've done a lot of work already - if you prefer we can ship the version that works with `--to` and do a lighter follow-up on `--sync`.

JohnPaton commented:
I'd rather just get it right, especially if `--sync` is the most common use case 😬  Indeed looks like my env was borked, recreating did the trick.

mwouts commented:
Hi @JohnPaton , well `--sync` first finds out which are the files paired with the current notebook. This information may come from the notebook metadata (`jupytext.formats`) or from the `jupytext.toml` configuration file.

Then, it determines which file is the most recent in this set, and then it updates the inputs cells of the other files with these most recent inputs.

In other words, if the `.ipynb` file was last modified, then it will update the `.py` file with the inputs cells from the `.ipynb` file. And if the `.py` file was last modified, the input cells of the `.ipynb` files are updated, using the content from the `.py` file (and outputs in the `.ipynb` are preserved, as much as possible).

The `--sync` mode is an attempt to replicate the Jupytext plugin (contents manager) for Jupyter at the command line. For an introduction to that one, you can have a look at this 12 minute presentation: [![](https://img.shields.io/badge/YouTube-JupyterCon%202020-red.svg)](https://www.youtube.com/watch?v=SDYdeVfMh48)

JohnPaton commented:
Almost there I think. The one last problem is that the hook seems to be updating both files in `--sync` mode (I was expecting only the older file to be updated). Because both files are being updated, the hook doesn't know which set of updates to take, so it fails... thoughts? 

```
jupytext.................................................................Failed
- hook id: jupytext
- files were modified by this hook

[jupytext] Ignoring unmatched input path .pre-commit-config.yaml
[jupytext] Reading test.ipynb in format ipynb
[jupytext] Updating 'test.py'
[jupytext] Reading test.py in format py
[jupytext] Loading 'test.ipynb'
[jupytext] Updating 'test.ipynb'
[jupytext] Updating 'test.py'

[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
```

mwouts commented:
Thank you @JohnPaton , you've done a hard work here!

Well you're right, when the `.py` file is more recent than the `.ipynb` file, `--sync` updates the `.ipynb` file and then updates (the timestamp) of the `.py` file.

Now I'll try to explain the reason for that... it is rooted in the other usage of Jupytext - when Jupytext is used within Jupyter. In that context, the user may edit either the `.py` file or the `.ipynb` file. We always take the `.py` file as the source for inputs, and we **throw an error** (we ask the user to delete one of the two files) if the `.py` file is older than the `.ipynb` file. This works (and is compatible with `--sync`, that was developed later on) as Jupytext always writes the `.ipynb` file before the `.py` version.

Maybe my preferred way to solve this would be, thus, to find a way to update the timestamp of the `.py` file (currently done at https://github.com/mwouts/jupytext/blob/master/jupytext/cli.py#L691-L693 by actually writing the file) that does not attract the attention of `pre-commit`. Do you think this is an option?

Another option is to assume that users of the pre-commit hook won't be using Jupytext's contents manager, and to skip the update of the unmodified file.

Or maybe the problem will exist anyway (`.ipynb` more recent than `.py` especially on large files) when we pull a commit from a remote? If so, maybe we could skip the update of the unmodified file, and use a post-commit hook to finally update the (timestamp) of the `.py` file to make it more recent than the `.ipynb`?

@Skylion007 , did that an issue appear in the local version of the hook? Do you have any advice on how to address this? Thanks

Skylion007 commented:
Yeah, to be honest the way I solve this is by deleting one of the files and
then rerunning it. It's not elegant, but it works. Yeah this is one of the
biggest issues with using it as a pre-commit hook currently. Updating the
timestamps of both files might be the way to go here. We can safely do that
now that we require_serial: True in our JuPyText hook.

We don't need to actually write the file btw, we can set the date modified
by using os.utime https://docs.python.org/3/library/os.html#os.utime

I prefer this solution as it still satisfies the JuPyText contents manager
and frankly both files should be updated when one is modified (even if the
update is a NOOP). This behavior could be changed by a flag (we may want to
group all these behaviors under a master flag pre-commit mode).

On Mon, Jan 18, 2021 at 6:16 PM Marc Wouts <notifications@github.com> wrote:

> Thank you @JohnPaton <https://github.com/JohnPaton> , you've done a hard
> work here!
>
> Well you're right, when the .py file is more recent than the .ipynb file,
> --sync updates the .ipynb file and then updates (the timestamp) of the .py
> file.
>
> Now I'll try to explain the reason for that... it is rooted in the other
> usage of Jupytext - when Jupytext is used within Jupyter. In that context,
> the user may edit either the .py file or the .ipynb file. We always take
> the .py file as the source for inputs, and we *throw an error* (we ask
> the user to delete one of the two files) if the .py file is older than
> the .ipynb file. This works (and is compatible with --sync, that was
> developed later on) as Jupytext always writes the .ipynb file before the
> .py version.
>
> Maybe my preferred way to solve this would be, thus, to find a way to
> update the timestamp of the .py file (currently done at
> https://github.com/mwouts/jupytext/blob/master/jupytext/cli.py#L691-L693
> by actually writing the file) that does not attract the attention of
> pre-commit. Do you think this is an option?
>
> Another option is to assume that users of the pre-commit hook won't be
> using Jupytext's contents manager, and to skip the update of the unmodified
> file.
>
> Or maybe the problem will exist anyway (.ipynb more recent than .py
> especially on large files) when we pull a commit from a remote? If so,
> maybe we could skip the update of the unmodified file, and use a
> post-commit hook to finally update the (timestamp) of the .py file to
> make it more recent than the .ipynb?
>
> @Skylion007 <https://github.com/Skylion007> , did that an issue appear in
> the local version of the hook? Do you have any advice on how to address
> this? Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mwouts/jupytext/pull/698#issuecomment-762509399>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPVMX2PHQO3KGFFZ765P4TS2S6LFANCNFSM4VOSIZLQ>
> .
>

mwouts commented:
> We don't need to actually write the file btw, we can set the date modified by using os.utime

Yes! I'll try to do that tonight (don't rewrite the `.py` file but just update the last modification time). I think it should be the default mode in `--sync` - no need for something specific to the pre-commit mode here, hopefully. But... will that work with pre-commit? How does pre-commit detect that the file has changed?

JohnPaton commented:
If it was only timestamps, it would notify that the timestamp is being updated, right? I don't see any
```python
"[jupytext] Sync timestamp of '{}'".format(nb_file)
```
line in my output

Skylion007 commented:
I haven't checked, but I believe pre-commit uses git to see which files don't match the index. It looks like git doesn't store any metadata about the file like ACLs or modified time so that might work: https://stackoverflow.com/a/2179825

mwouts commented:
Hi @JohnPaton , maybe this commit https://github.com/mwouts/jupytext/commit/22d6961572f6367945189d986f813ccd701f6084 can help? It generalizes the timestamp update (already there for the `--to` mode) to the `--sync` mode.

JohnPaton commented:
I'm afraid that wasn't enough either, it seems like both files are still being modified
```
jupytext.................................................................Failed
- hook id: jupytext
- files were modified by this hook

[jupytext] Ignoring unmatched input path .pre-commit-config.yaml
[jupytext] Reading test.ipynb in format ipynb
[jupytext] Updating 'test.py'
[jupytext] Reading test.py in format py
[jupytext] Loading 'test.ipynb'
[jupytext] Updating 'test.ipynb'
[jupytext] Sync timestamp of 'test.py'

[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
```

Not sure if this matters: `pre-commit` passes the files individually, iiuc, so it will call `jupytext --sync test.ipynb` and then `jupytext --sync test.py`

mwouts commented:
Hi @JohnPaton , yes you're right the issue is with the fact that `jupytext --sync` is called twice. On the second call the `.py` file is more recent so the `.ipynb` file is updated.

Here are a few workarounds that I am thinking of...
1. run the hook only on `.ipynb` files
2. call jupytext just once on the set of all files (and change `jupytext/cli.py` to run `jupytext` exactly once per paired notebook)
3. update `jupytext/cli.py` again, and skip the calls to 'write' if we see that the updated version matches the current file

I think that 1. is clearly the simple approach. And FYI I might also consider 3 because I think it will be required in case of triple-paired notebooks (`.ipynb` + `.md` + `.py`) (but be reassured, we will keep that for later on)

JohnPaton commented:
Alright, let's go with option 1 for now then! Then we can get this bad boy merged 🚀

mwouts commented:
Hi @JohnPaton , @Skylion007 , we're getting ready!

I am done with the review and I'd like to discuss a few details:
1. The example with `--to` in the doc needs to be updated to match the one actually tested (use e.g. https://github.com/mwouts/jupytext/commit/07b323b92da78d24b7f6c0887adc8f9d30607f8f)
2. Maybe we should temporarily remove the `--pipe` example from the documentation - at least I can't get it to work in the tests for now. I think option 3 from my previous comment could help here, but I'd prefer to re-add the example later on when it is ready.
3. How do you want to see this merged? Should I squash or rebase? If we squash, are you both OK for giving the full authorship to John (and I'll cite you both in the changelog)? Sorry for asking, it is the first time I see a PR with so many commits and contributors :smile: 
4. Github says that there is a conflict with the base branch, is it easy for you to solve? (Well if we decide to squash, don't bother with that, as I should be able to do it locally)

Skylion007 commented:
@mwouts Squashing commits will preserve authorship if done correctly (via the co-author tag in the commit description).

>> update jupytext/cli.py again, and skip the calls to 'write' if we see that the updated version matches the current file
I'd much prefer 3 in the long term. It's sane behavior.

mwouts commented:
Done! Thank you so much @JohnPaton and @Skylion007 . I am so happy to see this merged! What an impressive work!

I will create a follow-up issue to keep track of the few improvements that we have identified above, and also try to make progress on the blog post... I'll keep you posted!
